### PR TITLE
crimson/osd/pg_backend: fix rollback on whiteouts

### PR DESCRIPTION
**Pull request by xxhdx1985126**
1. There exists circumstances in which the dest-oid of rollback doesn't resolve to a valid clone
2. We might need to whiteout an object when rolling it back

Signed-off-by: Xuehan Xu <xuxuehan@qianxin.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by xxhdx1985126** [comment]
This pr is based #53306

**Comment by xxhdx1985126** [comment]
jenkins test api

**Comment by xxhdx1985126** [comment]
jenkins test windows

**Comment by Matan-B** [comment]
> This pr is based #53306

Let's merge this first, I'll schedule a run.

File: src/crimson/osd/pg_backend.cc [hunk 1]
```diff
@@ -833,12 +835,14 @@ PGBackend::rollback_iertr::future<> PGBackend::rollback(
...
-    [this, &os, &snapid, &txn, &delta_stats] {
+    [this, &os, &snapid, &txn, &delta_stats, &snapc, &ss] {
       logger().debug("PGBackend::rollback: deleting head on {}"
                      " with snap_id of {}"
                      " because got ENOENT|whiteout on obc lookup",
                      os.oi.soid, snapid);
-      return remove(os, txn, delta_stats, false);
+      bool whiteout = !ss.clones.empty() ||
+	(!snapc.snaps.empty() && snapc.snaps[0] > ss.seq);
+      return remove(os, txn, delta_stats, whiteout);
```
Comment (Matan-B) [inline src/crimson/osd/pg_backend.cc:845]: nit: Can this be moved to an helper method?
both CEPH_OSD_OP_DELETE and CEPH_OSD_OP_ROLLBACK re-use this logic.

Comment (Matan-B) [inline src/crimson/osd/pg_backend.cc:845]: nit: Can this be moved to an helper method?
both CEPH_OSD_OP_DELETE and CEPH_OSD_OP_ROLLBACK re-use this logic.

**Review by Matan-B — APPROVED** [review]
LGTM (one nit), let's merge #53306 first, I'll schedule a run.

File: src/crimson/osd/pg_backend.cc (hunk 1 above)
Comment (xxhdx1985126) [reply src/crimson/osd/pg_backend.cc]: Done.

File: src/osd/osd_types.cc [hunk 2]
```diff
@@ -5965,6 +5965,13 @@ SnapSet SnapSet::get_filtered(const pg_pool_t &pinfo) const
   return ss;
 }
 
+#ifdef WITH_SEASTAR
+bool should_whiteout(const SnapSet &ss, const SnapContext &client_snapc) {
```
Comment (Matan-B) [inline src/osd/osd_types.cc:5969]: nit: `inline static const bool`

File: src/osd/osd_types.cc [hunk 3]
```diff
@@ -5965,6 +5965,13 @@ SnapSet SnapSet::get_filtered(const pg_pool_t &pinfo) const
   return ss;
 }
 
+#ifdef WITH_SEASTAR
```
Comment (Matan-B) [inline src/osd/osd_types.cc:5968]: Note: This can probably be used also by classic (In a separate PR), so it's worth the change to common files.

**Comment by Matan-B** [comment]
https://github.com/ceph/ceph/pull/53306#issuecomment-1824369723

**Comment by athanatos** [comment]
jenkins test windows

**Comment by athanatos** [comment]
Seems to trigger

```
BUG 2023-12-15 22:52:50,740 [shard 2] osd -  pg_epoch 65 pg[41.2( v 44'3 (0'0,44'3] local-lis/les=64/65 n=2 ec=35/35 lis/c=64/35 les/c/f=65/41/0 sis=64) [0] r=0 lpr=64 pi=[35,64)/1 crt=44'4 lcod 41'1 mlcod 0'0 active+undersized+degraded  ObjectContextLoader::get_or_load_obc: cache hit on 41:42a55766:test-rados-api-smithi002-84725-7::foo:head
DEBUG 2023-12-15 22:52:50,740 [shard 1] osd -  pg_epoch 65 pg[49.11( empty local-lis/les=64/65 n=0 ec=45/45 lis/c=0/0 les/c/f=0/0/0 sis=64) [0] r=0 lpr=64 pi=[45,64)/1 crt=0'0 mlcod 0'0 creating+activating+undersized PeeringState::update_calc_stats degraded 0
DEBUG 2023-12-15 22:52:50,740 [shard 1] osd -  pg_epoch 65 pg[49.11( empty local-lis/les=64/65 n=0 ec=45/45 lis/c=0/0 les/c/f=0/0/0 sis=64) [0] r=0 lpr=64 pi=[45,64)/1 crt=0'0 mlcod 0'0 creating+activating+undersized PeeringState::update_calc_stats misplaced 0
DEBUG 2023-12-15 22:52:50,740 [shard 2] osd - resolve_oid oid.snap=2,head snapset.seq=3
DEBUG 2023-12-15 22:52:50,740 [shard 0] osd - peering_event(id=15879, detail=PeeringEvent(from=? pgid=52.17 sent=63 requested=63 evt=epoch_sent: 63 epoch_requested: 63 NullEvt +create_info)): got map 65, entering get_pg_mapping
DEBUG 2023-12-15 22:52:50,740 [shard 2] osd - resolve_oid {3: [3]} does not contain 2 -- DNE
DEBUG 2023-12-15 22:52:50,740 [shard 1] osd -  pg_epoch 65 pg[49.11( empty local-lis/les=64/65 n=0 ec=45/45 lis/c=0/0 les/c/f=0/0/0 sis=64) [0] r=0 lpr=64 pi=[45,64)/1 crt=0'0 mlcod 0'0 active+undersized state<Started/Primary/Active>:  mnow 120.414093018s >= prior_readable_until_ub 0.000000000s
INFO  2023-12-15 22:52:50,740 [shard 1] osd -  pg_epoch 65 pg[49.11( empty local-lis/les=64/65 n=0 ec=45/45 lis/c=0/0 les/c/f=0/0/0 sis=64) [0] r=0 lpr=64 pi=[45,64)/1 crt=0'0 mlcod 0'0 active+undersized state<Started/Primary/Active>: react AllReplicasActivated Activating complete
DEBUG 2023-12-15 22:52:50,740 [shard 0] osd - peering_event(id=15877, detail=PeeringEvent(from=? pgid=43.b sent=64 requested=64 evt=epoch_sent: 64 epoch_requested: 64 NullEvt +create_info)): have_pg
DEBUG 2023-12-15 22:52:50,740 [shard 0] osd - peering_event(id=15877, detail=PeeringEvent(from=? pgid=43.b sent=64 requested=64 evt=epoch_sent: 64 epoch_requested: 64 NullEvt +create_info)): have_pg
DEBUG 2023-12-15 22:52:50,740 [shard 1] osd -  pg_epoch 65 pg[49.11( empty local-lis/les=64/65 n=0 ec=45/45 lis/c=64/0 les/c/f=65/0/0 sis=64) [0] r=0 lpr=64 pi=[45,64)/1 crt=0'0 mlcod 0'0 active+undersized PeeringState::share_pg_info share_pg_info
ERROR 2023-12-15 22:52:50,740 [shard 2] none - /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.0.0-201-gcd750a7f/rpm/el9/BUILD/ceph-19.0.0-201-gcd750a7f/src/crimson/osd/osd_operations/client_request_common.cc:43 : In function 'crimson::osd::CommonClientRequest::recover_missings(Ref<crimson::osd::PG>&, const hobject
_t&, std::vector<snapid_t>&&)::<lambda(auto:179&)> mutable::<lambda()> mutable::<lambda(auto:180, auto:181)> mutable [with auto:180 = boost::intrusive_ptr<crimson::osd::ObjectContext>; auto:181 = boost::intrusive_ptr<crimson::osd::ObjectContext>]', ceph_assert(%s)
head->get_head_ss().clones.empty()
```

in my testing branch wip-sjust-crimson-testing-2023-12-15 in suite test

http://qa-proxy.ceph.com/teuthology/sjust-2023-12-15_22:32:12-crimson-rados-wip-sjust-crimson-testing-2023-12-15-distro-default-smithi/7493332/

File: src/crimson/osd/osd_operations/client_request_common.cc [hunk 4]
```diff
@@ -11,8 +11,60 @@ namespace {
...
+  return seastar::do_with(
+    std::move(snaps),
+    [pg, soid, fut=std::move(fut)](auto &snaps) mutable {
+    return fut.then_interruptible([&snaps, pg, soid]() mutable {
+      return pg->obc_loader.with_obc<RWState::RWREAD>(
+        soid.get_head(),
+        [&snaps, pg, soid](auto head, auto) mutable {
+        auto oid = resolve_oid(head->get_head_ss(), soid);
+        if (!oid) {
+          ceph_assert(head->get_head_ss().clones.empty());
```
Comment (athanatos) [inline src/crimson/osd/osd_operations/client_request_common.cc:43]: This assert is wrong -- if the operation is a read on a snapshot, there's no reason to assume that the snapshot maps to a point where the object existed.

**Comment by athanatos** [comment]
See the top two commits on https://github.com/athanatos/ceph/tree/sjust/wip-crimson-2023-12-13-fixes

File: src/crimson/osd/osd_operations/client_request_common.cc (hunk 4 above)
Comment (xxhdx1985126) [reply src/crimson/osd/osd_operations/client_request_common.cc]: Yeah, corrected.

**Comment by Matan-B** [comment]
jenkins test make check

**Comment by Matan-B** [comment]
https://pulpito.ceph.com/matan-2023-12-18_19:29:45-crimson-rados-wip-matanb-crimson-testing-misc-distro-crimson-smithi/ 

Only test failing is `rados_api_tests` although is was failing for quite a while now (with osd crashes). This failure seems different. Can be seen by grepping for `echo 'error in ` in the [log](http://qa-proxy.ceph.com/teuthology/matan-2023-12-18_19:29:45-crimson-rados-wip-matanb-crimson-testing-misc-distro-crimson-smithi/7495579/teuthology.log).

Iv'e scheduled this test 5 times to further debug:
https://pulpito.ceph.com/matan-2023-12-19_08:15:26-crimson-rados-wip-matanb-crimson-testing-misc-distro-crimson-smithi/

I'm ok with merging this based on the run attached and addressing the failure seperatly.

**Comment by Matan-B** [comment]
> Only test failing is `rados_api_tests` although is was failing for quite a while now (with osd crashes). This failure seems different. Can be seen by grepping for `echo 'error in ` in the [log](http://qa-proxy.ceph.com/teuthology/matan-2023-12-18_19:29:45-crimson-rados-wip-matanb-crimson-testing-misc-distro-crimson-smithi/7495579/teuthology.log).

Ah, found the issue. #52495 was merged and we used the qa dir with the test additions but without the corresponding code in the build. Attempting again with qa dir before the merge.

**Comment by Matan-B** [comment]
Merging based on:
https://pulpito.ceph.com/matan-2023-12-18_19:29:45-crimson-rados-wip-matanb-crimson-testing-misc-distro-crimson-smithi/

rados_api_tests rerun (with correct suite):
https://pulpito.ceph.com/matan-2023-12-19_09:17:50-crimson-rados-wip-matanb-crimson-testing-misc-distro-crimson-smithi/

Build:
https://shaman.ceph.com/builds/ceph/wip-matanb-crimson-testing-misc/

Note: the code adds function to the common code `osd_types.h`. However, the function has no callers in classical code. The build should be sufficient for merging. Moreover, the helper function is also usable in classical code (future PR).

**Pull request merged**